In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df = pd.read_csv('BL-Flickr-Images-Book.csv',sep=',',engine='python')

In [3]:
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [4]:
df.dtypes

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 15 columns):
Identifier                8287 non-null int64
Edition Statement         773 non-null object
Place of Publication      8287 non-null object
Date of Publication       8106 non-null object
Publisher                 4092 non-null object
Title                     8287 non-null object
Author                    6509 non-null object
Contributors              8287 non-null object
Corporate Author          0 non-null float64
Corporate Contributors    0 non-null float64
Former owner              1 non-null object
Engraver                  0 non-null float64
Issuance type             8287 non-null object
Flickr URL                8287 non-null object
Shelfmarks                8287 non-null object
dtypes: float64(3), int64(1), object(11)
memory usage: 971.2+ KB


In [6]:
df['Corporate Author'].isnull().value_counts()

True    8287
Name: Corporate Author, dtype: int64

In [7]:
df.drop(['Corporate Author','Edition Statement','Corporate Contributors','Former owner','Engraver','Shelfmarks'], inplace=True, axis=1)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 9 columns):
Identifier              8287 non-null int64
Place of Publication    8287 non-null object
Date of Publication     8106 non-null object
Publisher               4092 non-null object
Title                   8287 non-null object
Author                  6509 non-null object
Contributors            8287 non-null object
Issuance type           8287 non-null object
Flickr URL              8287 non-null object
dtypes: int64(1), object(8)
memory usage: 582.8+ KB


In [9]:
df.head(10)

Identifier                 Place of Publication Date of Publication  \
0         206                               London         1879 [1878]   
1         216             London; Virtue & Yorston                1868   
2         218                               London                1869   
3         472                               London                1851   
4         480                               London                1857   
5         481                               London                1875   
6         519                               London                1872   
7         667  pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
8         874                              London]                1676   
9        1143                               London                1679   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   
5      William Macintosh  [The World in which I live, and my place in it...   
6             The Author  Lagonells. By the author of Darmayne (F. E. A....   
7                    NaN  The Coming of Spring, and other poems. By J. A...   
8                    NaN  A Warning to the inhabitants of England, and L...   
9                    NaN  A Satyr against Vertue. (A poem: supposed to b...   

          Author                                  Contributors Issuance type  \
0          A. A.                               FORBES, Walter.   monographic   
1      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3      A., E. S.                   Appleyard, Ernest Silvanus.   monographic   
4      A., E. S.                           BROOME, John Henry.   monographic   
5      A., E. S.                           BROOME, John Henry.   monographic   
6      A., F. E.                       ASHLEY, Florence Emily.   monographic   
7  A., J.|A., J.                 ANDREWS, J. - Writer of Verse   monographic   
8         Remaʿ.                                  ADAMS, Mary.   monographic   
9         A., T.                                 OLDHAM, John.   monographic   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...  
5  http://www.flickr.com/photos/britishlibrary/ta...  
6  http://www.flickr.com/photos/britishlibrary/ta...  
7  http://www.flickr.com/photos/britishlibrary/ta...  
8  http://www.flickr.com/photos/britishlibrary/ta...  
9  http://www.flickr.com/photos/britishlibrary/ta...

In [10]:
df['Date of Publication'].value_counts()

1897               157
1896               150
1893               130
1892               127
1898               125
1895               124
1899               120
1891               119
1894               118
1890               115
1876               113
1889               113
1888               105
1880               103
1887                99
1884                96
1886                93
1878                93
1869                92
1881                91
1882                89
1875                87
1874                87
1885                83
1873                82
1864                81
1866                81
1879                80
1883                80
1877                79
                  ... 
[1817]               1
[1883?]              1
1866-68              1
1855 [1853-55]       1
[1789.]              1
1819-[20]            1
1824, 1825           1
1852]                1
[1892?]              1
[1871, 72]           1
[1817.]              1
1858 [1851-58]       1
1864 [1863]

In [11]:
def date_change(value):
    if not pd.isnull(value):
        dt = value[:6]
        if '[' in dt[0]:
            date = dt[1:5]
            return date
        else:
            date = dt[:4]
            return date
        

In [12]:
df['Date of Publication'] = df['Date of Publication'].apply(date_change)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 9 columns):
Identifier              8287 non-null int64
Place of Publication    8287 non-null object
Date of Publication     8106 non-null object
Publisher               4092 non-null object
Title                   8287 non-null object
Author                  6509 non-null object
Contributors            8287 non-null object
Issuance type           8287 non-null object
Flickr URL              8287 non-null object
dtypes: int64(1), object(8)
memory usage: 582.8+ KB


In [14]:
df['Date of Publication'].unique()

array(['1879', '1868', '1869', '1851', '1857', '1875', '1872', None,
       '1676', '1679', '1802', '1859', '1888', '1839', '1897', '1865',
       '1860', '1873', '1866', '1899', '1814', '1820', '1800', '1847',
       '1893', '1805', '1837', '1896', '1898', '1892', '1894', '1885',
       '1846', '1817', '1816', '1833', '1804', '1777', '1799', '1827',
       '1853', '1874', '1790', '1883', '1795', '1877', '1886', '1834',
       '1852', '1828', '1876', '1758', '1880', '1823', '1887', '1825',
       '1850', '1810', '1889', '1861', '1858', '1878', '1821', '1891',
       '1808', '1849', '1724', '1772', '1812', '1835', '1867', '1830',
       '1841', '1884', '1863', '1848', '1845', '1807', '1864', '1822',
       '1871', '1829', '1824', '1856', '1803', '1818', '1881', '1838',
       '1720', '1855', '1747', '1748', '1704', '1762', '1694', '1890',
       '1895', '1870', '1809', '1760', '1862', '1844', '1813', '1787',
       '1794', '1710', '1785', '1801', '1882', '1840', '1819', '1651',
       '

In [15]:
def date_cln(date):
    try:
        return np.int64(date)
    except:
        return np.nan

In [16]:
df['Date of Publication'] = df['Date of Publication'].apply(date_cln)

In [17]:
df['Date of Publication'].unique()

array([ 1879.,  1868.,  1869.,  1851.,  1857.,  1875.,  1872.,    nan,
        1676.,  1679.,  1802.,  1859.,  1888.,  1839.,  1897.,  1865.,
        1860.,  1873.,  1866.,  1899.,  1814.,  1820.,  1800.,  1847.,
        1893.,  1805.,  1837.,  1896.,  1898.,  1892.,  1894.,  1885.,
        1846.,  1817.,  1816.,  1833.,  1804.,  1777.,  1799.,  1827.,
        1853.,  1874.,  1790.,  1883.,  1795.,  1877.,  1886.,  1834.,
        1852.,  1828.,  1876.,  1758.,  1880.,  1823.,  1887.,  1825.,
        1850.,  1810.,  1889.,  1861.,  1858.,  1878.,  1821.,  1891.,
        1808.,  1849.,  1724.,  1772.,  1812.,  1835.,  1867.,  1830.,
        1841.,  1884.,  1863.,  1848.,  1845.,  1807.,  1864.,  1822.,
        1871.,  1829.,  1824.,  1856.,  1803.,  1818.,  1881.,  1838.,
        1720.,  1855.,  1747.,  1748.,  1704.,  1762.,  1694.,  1890.,
        1895.,  1870.,  1809.,  1760.,  1862.,  1844.,  1813.,  1787.,
        1794.,  1710.,  1785.,  1801.,  1882.,  1840.,  1819.,  1651.,
      

In [18]:
df['Date of Publication'].dtype

dtype('float64')

In [19]:
df['Place of Publication'].value_counts()

London                                            3868
Paris                                              479
Edinburgh                                          208
New York                                           177
Leipzig                                            119
Philadelphia                                        89
Berlin                                              70
Boston [Mass.]                                      52
Dublin                                              48
Glasgow                                             45
Oxford                                              44
Boston                                              41
Wien                                                38
Madrid                                              33
Stockholm                                           33
Bruxelles                                           28
Cambridge                                           26
Amsterdam                                           25
Firenze   

In [20]:
df.head(50)

Identifier                 Place of Publication  Date of Publication  \
0          206                               London               1879.0   
1          216             London; Virtue & Yorston               1868.0   
2          218                               London               1869.0   
3          472                               London               1851.0   
4          480                               London               1857.0   
5          481                               London               1875.0   
6          519                               London               1872.0   
7          667  pp. 40. G. Bryan & Co: Oxford, 1898                  NaN   
8          874                              London]               1676.0   
9         1143                               London               1679.0   
10        1280                             Coventry               1802.0   
11        1808                          Christiania               1859.0   
12        1905                              Firenze               1888.0   
13        1929                            Amsterdam               1839.0   
14        2836                               Savona               1897.0   
15        2854                               London               1865.0   
16        2956                                Paris               1860.0   
17        2957                                Paris               1873.0   
18        3017                          Puerto-Rico               1866.0   
19        3131                             New York               1899.0   
20        4598                                 Hull               1814.0   
21        4884                               London               1820.0   
22        4976                               Oxonii               1800.0   
23        5382                               London               1847.0   
24        5385                               London               1897.0   
25        5389                               London               1897.0   
26        5432                               Milano               1893.0   
27        6036                               London               1805.0   
28        6821                             Aberdeen               1837.0   
29        7521                                 Wien               1896.0   
30        7630                             Abingdon               1898.0   
31        8239                               Quebec               1899.0   
32        8435                               London               1892.0   
33        8440                               London               1894.0   
34       11361                              Leipzig               1894.0   
35       11852   pp. 40. W. Cann: Plymouth, [1876?]                  NaN   
36       13074                               London               1899.0   
37       13364                               London               1885.0   
38       14466                               London               1860.0   
39       14703                                 1845               1846.0   
40       15141                               London               1893.0   
41       15146                               London               1894.0   
42       16173                                  enk               1817.0   
43       16543                               London               1816.0   
44       16544                               Boston               1817.0   
45       17602                               London               1897.0   
46       17752                              Upsaliæ               1833.0   
47       18173                               London               1804.0   
48       18218                            Edinburgh               1859.0   
49       18347                              Yverdon               1777.0   

                                            Publisher  \
0                                    S. Tinsley & Co.   
1            

# Check for names that start with pp

In [21]:
pub_list = []
for x in df.ix[:,'Place of Publication']:
    if x.startswith('pp.'):
        pub_list.append(x)

In [22]:
pub_list

['pp. 40. G. Bryan & Co: Oxford, 1898',
 'pp. 40. W. Cann: Plymouth, [1876?]',
 'pp. vi. 216. Hatchards: London, 1888',
 'pp. 64. W. Cann: Plymouth, [1870?]',
 'pp. 95. Erskine Macdonald: London, 1921',
 'pp. xi. 59. J. Parker & Co.: Oxford, [1900.]',
 'pp. xii. 59. Thomas Burleigh: London, 1900',
 'pp. 92. Heath Cranton: London, 1919',
 'pp. x. 429. John Murray: London, 1894',
 'pp. 31. S. Harper & Sons: Barnstaple, 1899',
 'pp. 281. Hatchards: London, [1889.]',
 'pp. 128. R. Aickin & Co.: Belfast, 1896',
 'pp. 64. W. & J. Kennedy: Hawick, 1899',
 'pp. 359. Thomas Nelson & Sons: London, 1889 [1888]',
 'pp. 349. Griffith, Farran, Okeden & Welsh: London, [1888.]',
 'pp. 59. Blackwell: Oxford, [1919.]',
 'pp. 110. John Lane: London, New York, 1918',
 'pp. 221. George Routledge & Sons: London, 1887',
 'pp. 152. Москва, 1898',
 'pp. vi. 250. W. Blackwood: Edinburgh, 1889',
 'pp. 332. Hurst & Blackett: London; Bolton [printed], 1889 [1888.]',
 'pp. vi. 344. Chapman & Hall: London, 1887',
 '

In [23]:
len(pub_list)

113

# Split name that start with 'pp' at ':' and extract name of the place after ':'

In [24]:
def extract_place(place):
    if place.startswith('pp.'):
        try:
            part_plc = place.split(':')[1].strip()
            if ',' in part_plc:
                place = part_plc[:part_plc.find(',')]
            elif ';' in part_plc:
                place = part_plc[:part_plc.find(',')]
            return place
        except:
            return np.nan
    else:
        if 'London' in place:
            place = 'London'
        return place
        

In [25]:
df['Place of Publication'] = df['Place of Publication'].apply(extract_place)

In [26]:
df['Place of Publication']

0                    London
1                    London
2                    London
3                    London
4                    London
5                    London
6                    London
7                    Oxford
8                    London
9                    London
10                 Coventry
11              Christiania
12                  Firenze
13                Amsterdam
14                   Savona
15                   London
16                    Paris
17                    Paris
18              Puerto-Rico
19                 New York
20                     Hull
21                   London
22                   Oxonii
23                   London
24                   London
25                   London
26                   Milano
27                   London
28                 Aberdeen
29                     Wien
               ...         
8257                 London
8258                 London
8259                 London
8260                 London
8261                

# Check of other irregularities

In [27]:
for x in df.ix[:,'Place of Publication']:
    print(x)

London
London
London
London
London
London
London
Oxford
London
London
Coventry
Christiania
Firenze
Amsterdam
Savona
London
Paris
Paris
Puerto-Rico
New York
Hull
London
Oxonii
London
London
London
Milano
London
Aberdeen
Wien
Abingdon
Quebec
London
London
Leipzig
Plymouth
London
London
London
1845
London
London
enk
London
Boston
London
Upsaliæ
London
Edinburgh
Yverdon
Londini
St. Petersburg
Berlin
London
London
London
London
Sceaux
London
Coblenz
London
Leipzig
London
London
Carlstad, Stockholm
Carlstad
Boston
Paris
London
London
London
London
London
A. Millar
New York
Allahabad]
London
Stuttgart
London
D. Appleton & Co
London
Санктпетербургъ
London
Paisley
London
London
London
London
London
London
London
London
London
Paris
London
Edinburgh
Napoli
London
Ἐν Κερκυρᾳ
London
London
Plymouth
London
Trento
Amsterdam
Heilbronn
Torino
w Krakowie
w Krakowie
London
London
London
London
London
London
London
Boston
London
London
London
London
Edinburgh
Edinburgh
Cambridge, [Mass.]
Wien
London
Lo

Chicago
Chicago
Besançon
London
London
London
Perth
Paris
London
Leipzig
Leipzig
London
London
Cambridge
Plymouth
London
London
Kjøbenhavn
London
Stockholm
С.-Петербургъ
С.-Петербургъ
Göteborg
Dublin
Wien
Buenos Aires
Mexico
London
London
London
Paris
Leipzig
London
London
London
Paris
New York
Marburg
Paris
Munich
London
London
London
London
London
London
London
London
London
Bath
London
London
Birmingham
London
London
Boston [Mass.]
Lausanne
London
London
Edinburgh
Louvain
Nimes
Cambridge
Washington
London
Compiègne
New York
Nantes
Chur
London
London
London
Edinburgh
Paris
London
Exeter
London
London
London
London
London
London
London
London
London
London
London
London
London
London
Lisboa
London
Glasgow
London
München
Weimar
London
London
New York
Firenze
London
London
Newcastle-upon-Tyne
London
London
London
London
London
Stuttgart & Tübingen
Lipsiae
Lipsiæ
Lipsiae
Londini
Berolini
Paris
Lipsiae
Lugduni Batavorum
Monachi
Londini
Groningæ
Paris
London
Leeds
New York
London
Edin

In [28]:
df.head(25)

Identifier Place of Publication  Date of Publication  \
0          206               London               1879.0   
1          216               London               1868.0   
2          218               London               1869.0   
3          472               London               1851.0   
4          480               London               1857.0   
5          481               London               1875.0   
6          519               London               1872.0   
7          667               Oxford                  NaN   
8          874               London               1676.0   
9         1143               London               1679.0   
10        1280             Coventry               1802.0   
11        1808          Christiania               1859.0   
12        1905              Firenze               1888.0   
13        1929            Amsterdam               1839.0   
14        2836               Savona               1897.0   
15        2854               London               1865.0   
16        2956                Paris               1860.0   
17        2957                Paris               1873.0   
18        3017          Puerto-Rico               1866.0   
19        3131             New York               1899.0   
20        4598                 Hull               1814.0   
21        4884               London               1820.0   
22        4976               Oxonii               1800.0   
23        5382               London               1847.0   
24        5385               London               1897.0   

                Publisher                                              Title  \
0        S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1            Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2   Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3           James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4    Wertheim & Macintosh  [The World in which I live, and my place in it...   
5       William Macintosh  [The World in which I live, and my place in it...   
6              The Author  Lagonells. By the author of Darmayne (F. E. A....   
7                     NaN  The Coming of Spring, and other poems. By J. A...   
8                     NaN  A Warning to the inhabitants of England, and L...   
9                     NaN  A Satyr against Vertue. (A poem: supposed to b...   
10   Printed by J. Turner  An Account of the many and great Loans, Benefa...   
11                    NaN  Erindringer som Bidrag til Norges Historie fra...   
12                    NaN  Gli Studi storici in terra d'Otranto ... Framm...   
13                    NaN  De Aardbol. Magazijn van hedendaagsche land- e...   
14                    NaN  Cronache Savonesi dal 1500 al 1570 ... Accresc...   
15         E. Moxon & Co.  See-Saw; a novel ... Edited [or rather, writte...   
16                    NaN  Géodésie d'une partie de la Haute Éthiopie,...   
17                    NaN                                [With eleven maps.]   
18                    NaN  [Historia geográfica, civil y politica de la ...   
19              W. Abbatt  The Crisis of the Revolution, being the story ...   
20             The Author  Peace: a lyric poem. [With prefatory address b...   
21      J. Hatchard & Son  Abdallah; or, The Arabian Martyr: a Christian ...   
22         J. Cooke, etc.  [Abdollatiphi Historiæ Ægypti compendium. [Wit...   
23           Punch Office  The Comic History of England ... With ... colo...   
24  Bradbury, Agnew & Co.  [The comic history of England ... With twenty ...   

                                               Author  \
0                                               A. A.   
1                                           A., A. A.   
2                                           A., A. A.   
3                                           A., E. S.   
4                                           A., E. S.   
5                   

In [29]:
l = [x for x in df['Place of Publication'].unique()]
print(len(l))

1166


In [30]:
for i in l:
    print(i)

London
Oxford
Coventry
Christiania
Firenze
Amsterdam
Savona
Paris
Puerto-Rico
New York
Hull
Oxonii
Milano
Aberdeen
Wien
Abingdon
Quebec
Leipzig
Plymouth
1845
enk
Boston
Upsaliæ
Edinburgh
Yverdon
Londini
St. Petersburg
Berlin
Sceaux
Coblenz
Carlstad, Stockholm
Carlstad
A. Millar
Allahabad]
Stuttgart
D. Appleton & Co
Санктпетербургъ
Paisley
Napoli
Ἐν Κερκυρᾳ
Trento
Heilbronn
Torino
w Krakowie
Cambridge, [Mass.]
Bombay
Agen
Edinburgh; John Murray
Philadelphia
Haarlem
Albany
Kjøbenhavn
Haderslev
Copenhague
Kiel
Baltimore
S. Paulo
Londini Gothorum
Mamers
México
Mexico
Porto
Bruxelles
Macerata
Tours
New-York
Palermo
Santiago
Edinburgh]
Chelmsford
Barcelona
Glasgow
Inverness
Jedburgh
Westminster
Andover
Lewes
Deventer
Providence
Frankfurt
Roma
Pescia
Exeter
Malton
Great Yarmouth
Chur
Moulins
Madrid
Toulouse
Berolini
Lipsiae
Niort
Stockholm
Lecco
Carlisle
Dieppe
Pest
Dorchester
Jenae
Kristiania
Salisbury
Meriden, Conn
Paris; Moulins
Trévoux
München
Rouen
Adelaide
Perth
Montevideo
Bath
Bud

In [31]:
df.groupby('Publisher').count()

Identifier  \
Publisher                                                        
, 42. Nichols & Co.                                          1   
-152. D. H. Edwards                                          1   
-82. 1-298. E. P. for H. Seyle                               1   
1. E. Moxon & Co.                                            1   
1. Printed for the Author by W. Eyres                        1   
1. pp. 224. Lackington, Allen & Co.                          1   
1. pp. xxiii. 401. James Burns                               1   
10. G. G. & J. Robinson                                      1   
13. William Pearson                                          1   
135. 116. 41. J. Murray                                      1   
150. 115. 43. 4. Printed for the Editor                      1   
2 vols. Ward & Downey                                        1   
28. G. Odell                                                 1   
3 vols. Smith, Elder & Co.                                   1   
3. Privately printed                                         2   
344. Hurst & Blackett                                        1   
381. Macmillan & Co.                                         1   
4 no. pp. 104. William Strange                               1   
6. W. Owen                                                   1   
8. Reeves & Turner                                           1   
8. The Author                                                1   
96. Longman, Rees & Co.                                      1   
A spese dell’autore                                          1   
A. & C. Black                                               13   
A. & W. Galignani & Co.                                      1   
A. & W. Huke                                                 1   
A. & W. Mackenzie                                            1   
A. Bartholoman                                               1   
A. Bean                                                      1   
A. Bettesworth                                               1   
...                                                        ...   
pp. xxiii. 72. 617. G. Davidson                              1   
pp. xxvi                                                     1   
printed by Andrew Marschalk                                  1   
printed by Charles N. Dent                                   1   
printed by John Kemp                                         1   
printed by W. Thorne, and published by J. Callo...           1   
printed for John Ross, bookseller in Edinburgh               1   
printed for private circulation                              1   
published for the Proprietors by J. Murray, etc.             1   
viii. Morris Bros.                                           1   
x. Newsagents' Publishing Co.                                1   
xlviii. How & Parsons                                        1   
xviii. 4. Charles Haselden                                   1   
xviii. Longman                                               1   
xxiii. vii. 183. Blackwood & Sons                            1   
xxxiv. W. Blackwood & Sons                                   1   
“Albion” Press                                               1   
“Art Journal” Office                                         1   
“Brighton Gazette” Printing Co.                              1   
“Craven Herald”                                              1   
“Daily Herald” Office                                        1   
“Engineering” Office                                         1   
“Home Words” Publishing Office                               1   
“Judy” Office                                                3   
“Mirror”                                                     1   
“Moray Weekly News”                                          1   
“Northern Chronicle”                                         1   
“Sussex Advertiser” Office                                   1   
“Telegraph”               

# Group by 'Publisher' to see the top publisher's, here w.r.t 'Place of Publication'

In [32]:
df['Place of Publication'].groupby(df['Publisher']).count().sort_values(ascending=False)[:50]

Publisher
Macmillan & Co.                    85
London                             69
Privately printed                  53
Longmans & Co.                     52
Hurst & Blackett                   51
John Murray                        49
Smith, Elder & Co.                 47
Richard Bentley                    47
Sampson Low & Co.                  46
Chatto & Windus                    43
Chapman & Hall                     42
G. Routledge & Sons                30
W. Blackwood & Sons                27
R. Bentley & Son                   27
Printed for the Author             27
Printed for private circulation    25
Kegan Paul & Co.                   24
T. Fisher Unwin                    24
Longman & Co.                      24
Richard Bentley & Son              23
T. F. Unwin                        22
Cassell & Co.                      20
G. Philip & Son                    19
Harper & Bros.                     19
Ward, Lock & Co.                   18
J. Murray                          18
W.

# Check if the above no. 1 is true

In [33]:
df[df['Publisher']=='Macmillan & Co.']

Identifier Place of Publication  Date of Publication        Publisher  \
102        40021               London               1886.0  Macmillan & Co.   
210        79249               London               1876.0  Macmillan & Co.   
239        94484               London               1887.0  Macmillan & Co.   
306       124034               London               1886.0  Macmillan & Co.   
352       144509               London               1896.0  Macmillan & Co.   
353       144558               London               1897.0  Macmillan & Co.   
354       144597               London               1897.0  Macmillan & Co.   
355       144653                  enk               1895.0  Macmillan & Co.   
451       176559               London               1874.0  Macmillan & Co.   
496       201740               London               1870.0  Macmillan & Co.   
749       365514               London               1874.0  Macmillan & Co.   
858       419927               London               1896.0  Macmillan & Co.   
873       429755               London               1885.0  Macmillan & Co.   
967       474227               London               1882.0  Macmillan & Co.   
997       486800               London               1890.0  Macmillan & Co.   
1027      496315               London               1869.0  Macmillan & Co.   
1385      612625            Cambridge               1862.0  Macmillan & Co.   
1392      617597               London               1898.0  Macmillan & Co.   
1741      789138               London               1884.0  Macmillan & Co.   
1772      799739               London               1887.0  Macmillan & Co.   
1828      814437               London               1899.0  Macmillan & Co.   
1837      818824               London               1875.0  Macmillan & Co.   
1906      850914               London               1886.0  Macmillan & Co.   
2066      913657               London               1859.0  Macmillan & Co.   
2085      923498               London               1894.0  Macmillan & Co.   
2114      931072               London               1893.0  Macmillan & Co.   
2124      932267               London               1886.0  Macmillan & Co.   
2166      956292               London               1894.0  Macmillan & Co.   
2171      957050               London               1896.0  Macmillan & Co.   
2289     1006232               London               1896.0  Macmillan & Co.   
...          ...                  ...                  ...              ...   
4835     2390952               London               1895.0  Macmillan & Co.   
4836     2391015               London               1896.0  Macmillan & Co.   
4837     2391060               London               1897.0  Macmillan & Co.   
4838     2391161               London               1896.0  Macmillan & Co.   
4841     2391357               London               1895.0  Macmillan & Co.   
4842     2391411               London               1896.0  Macmillan & Co.   
4844     2391483               London               1897.0  Macmillan & Co.   
5111     2513349               London               1883.0  Macmillan & Co.   
5195     2550725               London               1895.0  Macmillan & Co.   
5268     2589981               London               1891.0  Macmillan & Co.   
5518     2703933               London               1890.0  Macmillan & Co.   
5526     2705275               London               1885.0  Macmillan & Co.   
5713     2798431               London               1898.0  Macmillan & Co.   
6136     3014995               London               1895.0  Macmillan & Co.   
6329     3108610               London               1880.0  Macmillan & Co.   
6527     3212144               London               1896.0  Macmillan & Co.   
6681     3287627               London               1898.0  Macmillan & Co.   
6878     3364363               London               1880.0  Macmillan & Co.   
7043     3440757               London               1889

# Group the whole data frame by publisher to see highest no. of authors with particular publishers

In [34]:
df.groupby(df['Publisher']).count().sort_values('Author',ascending=False)[:20]

Identifier  Place of Publication  \
Publisher                                                           
Macmillan & Co.                          85                    85   
London                                   69                    69   
Longmans & Co.                           52                    52   
John Murray                              49                    49   
Chatto & Windus                          43                    43   
Sampson Low & Co.                        46                    46   
Chapman & Hall                           42                    42   
Privately printed                        53                    53   
G. Routledge & Sons                      30                    30   
Richard Bentley                          47                    47   
T. Fisher Unwin                          24                    24   
Hurst & Blackett                         51                    51   
Kegan Paul & Co.                         24                    24   
T. F. Unwin                              22                    22   
G. Philip & Son                          19                    19   
R. Bentley & Son                         27                    27   
Smith, Elder & Co.                       47                    47   
Printed for private circulation          25                    25   
W. Blackwood & Sons                      27                    27   
Longman & Co.                            24                    24   

                                 Date of Publication  Title  Author  \
Publisher                                                             
Macmillan & Co.                                   85     85      68   
London                                            69     69      58   
Longmans & Co.                                    52     52      46   
John Murray                                       49     49      40   
Chatto & Windus                                   43     43      40   
Sampson Low & Co.                                 46     46      38   
Chapman & Hall                                    42     42      34   
Privately printed                                 53     53      33   
G. Routledge & Sons                               30     30      29   
Richard Bentley                                   47     47      25   
T. Fisher Unwin                                   24     24      22   
Hurst & Blackett                                  51     51      20   
Kegan Paul & Co.                                  24     24      20   
T. F. Unwin                                       22     22      20   
G. Philip & Son                                   19     19      19   
R. Bentley & Son                                  27     27      19   
Smith, Elder & Co.                                47     47      19   
Printed for private circulation                   25     25      18   
W. Blackwood & Sons                               27     27      18   
Longman & Co.                                     24     24      18   

                                 Contributors  Issuance type  Flickr URL  
Publisher                                                                 
Macmillan & Co.                            85             85          85  
London                                     69             69          69  
Longmans & Co.                             52             52          52  
John Murray                                49             49          49  
Chatto & Windus                            43             43          43  
Sampson Low & Co.                          46             46          46  
Chapman & Hall                             42             42          42  
Privately printed                          53             53          53  
G. Routledge & Sons                        30             30          30  
Richard Bentley                            47             47          47  
T. Fisher Unwin                            24      

# To find which Publishers have the highest Identifiers.

In [35]:
df.groupby(df['Publisher']).count().sort_values('Identifier',ascending=False)[:20]

Identifier  Place of Publication  \
Publisher                                                           
Macmillan & Co.                          85                    85   
London                                   69                    69   
Privately printed                        53                    53   
Longmans & Co.                           52                    52   
Hurst & Blackett                         51                    51   
John Murray                              49                    49   
Smith, Elder & Co.                       47                    47   
Richard Bentley                          47                    47   
Sampson Low & Co.                        46                    46   
Chatto & Windus                          43                    43   
Chapman & Hall                           42                    42   
G. Routledge & Sons                      30                    30   
R. Bentley & Son                         27                    27   
W. Blackwood & Sons                      27                    27   
Printed for the Author                   27                    27   
Printed for private circulation          25                    25   
Longman & Co.                            24                    24   
T. Fisher Unwin                          24                    24   
Kegan Paul & Co.                         24                    24   
Richard Bentley & Son                    23                    23   

                                 Date of Publication  Title  Author  \
Publisher                                                             
Macmillan & Co.                                   85     85      68   
London                                            69     69      58   
Privately printed                                 53     53      33   
Longmans & Co.                                    52     52      46   
Hurst & Blackett                                  51     51      20   
John Murray                                       49     49      40   
Smith, Elder & Co.                                47     47      19   
Richard Bentley                                   47     47      25   
Sampson Low & Co.                                 46     46      38   
Chatto & Windus                                   43     43      40   
Chapman & Hall                                    42     42      34   
G. Routledge & Sons                               30     30      29   
R. Bentley & Son                                  27     27      19   
W. Blackwood & Sons                               27     27      18   
Printed for the Author                            27     27      14   
Printed for private circulation                   25     25      18   
Longman & Co.                                     24     24      18   
T. Fisher Unwin                                   24     24      22   
Kegan Paul & Co.                                  24     24      20   
Richard Bentley & Son                             23     23      11   

                                 Contributors  Issuance type  Flickr URL  
Publisher                                                                 
Macmillan & Co.                            85             85          85  
London                                     69             69          69  
Privately printed                          53             53          53  
Longmans & Co.                             52             52          52  
Hurst & Blackett                           51             51          51  
John Murray                                49             49          49  
Smith, Elder & Co.                         47             47          47  
Richard Bentley                            47             47          47  
Sampson Low & Co.                          46             46          46  
Chatto & Windus                            43             43          43  
Chapman & Hall                             42      

# Cleaning up the nan's

In [36]:
df.columns

Index(['Identifier', 'Place of Publication', 'Date of Publication',
       'Publisher', 'Title', 'Author', 'Contributors', 'Issuance type',
       'Flickr URL'],
      dtype='object')

In [37]:
df[df['Place of Publication'].isnull()]

Identifier Place of Publication  Date of Publication Publisher  \
2160      954240                  NaN                  NaN       NaN   
4497     2213500                  NaN                  NaN       NaN   
4783     2370941                  NaN                  NaN       NaN   
5208     2556554                  NaN                  NaN       NaN   
5488     2688596                  NaN                  NaN       NaN   
5551     2718848                  NaN                  NaN       NaN   
6532     3212447                  NaN                  NaN       NaN   
6738     3317692                  NaN                  NaN       NaN   
7317     3578184                  NaN                  NaN       NaN   
7473     3641137                  NaN                  NaN       NaN   
7637     3761232                  NaN                  NaN       NaN   
8195     3995356                  NaN                  NaN       NaN   

                                                  Title  \
2160  Пугачевщина въ Сибири. Очеркъ по документамъ э...   
4497  Lohengrin Fifty Years After. [A poem.] By One ...   
4783  Записки Манштейна о Россіи, 1724-1744 гг. Замѣ...   
5208  The Pilgrims of Hope: a poem, etc. [With a pre...   
5488  Изслѣдованія и замѣтки Князя М. А. Оболенскаго...   
5551  Очеркъ трехнедѣльнаго похода Наполеона вротивъ...   
6532  Ezra. [In verse. Signed, E. S. G. S., i.e. E. ...   
6738  [The Lay of the Last Minstrel. By Sir Walter S...   
7317  Бернгардъ Таннеръ. Описаніе путешествія польск...   
7473  Floating Spars. [Verses. With a memoir of the ...   
7637  Исторія Вятскаго края съ древнихъ временъ до н...   
8195  Сочиненія Князя Курбскаго, какъ историческій м...   

                                                 Author  \
2160   DMITRIEV-MAMONOV, Aleksandr Ippolitovich - Count   
4497                                                NaN   
4783            MANSTEIN, Christoph Hermann von - Baron   
5208                                    Morris, William   
5488             OBOLENSKY, Mikhail Andreevich - Prince   
5551               ORLOV, Nikolai Aleksyeevich - Prince   
6532                                       S., E. S. G.   
6738                                Scott, Walter - Sir   
7317           TANNER, Bernhardus Leopoldus Franciscus.   
7473                          TINDALE, Thomas Plumpton.   
7637  VASIL'EV, S. - of Vyatka and BEKHTEREV (Nikola...   
8195                          YASINSKY, Anton Nikitich.   

                                           Contributors Issuance type  \
2160  DEKOLONG, Ivan Aleksandrovich.|PUGACHEV, Emel'...   monographic   
4497                            WESTON, Jessie Laidlay.   monographic   
4783                      PANIN, Petr Ivanovich - Count   monographic   
5208                              FORMAN, Harry Buxton.   monographic   
5488              KHILKOVA, Anna Mikhailovna - Princess   monographic   
5551                    KORF, Modest Andreevich - Baron   monographic   
6532                    SAUNDERS, Emily Susan Goulding.   monographic   
6738                                   HENKEL, Wilhelm.   monographic   
7317                         IVAKIN, Ivan Mikhailovich.   monographic   
7473                                          P., H. N.   monographic   
7637                      BEKHTEREV, Nikolai Pavlovich.   monographic   
8195              KURBSKY, Andrei Mikhailovich - Prince   monographic   

                                             Flickr URL  
2160  http://www.flickr.com/photos/britishlibrary/ta...  
4497  http://www.flickr.com/photos/britishlibrary/ta...  
4783  http://www.flickr.com/photos/britishlibrary/ta...  
5208  http://www.flickr.com/photos/britishlibrary/ta...  
5488  http://www.flickr.com/photos/britishlibrary/ta...  
5551  http://www.flickr.com/photos/britishlibrary/ta...  
6532  http://www.flickr.com/photos/britishlibrary/ta...  
6738  http://www.flickr.com/photos/britishlibrary/ta...  
7317  http://www.flickr.com/photos/britishlibrary/ta...  
747

In [38]:
for col in df.columns:
    print(df[col].isnull().value_counts())

False    8287
Name: Identifier, dtype: int64
False    8275
True       12
Name: Place of Publication, dtype: int64
False    8095
True      192
Name: Date of Publication, dtype: int64
True     4195
False    4092
Name: Publisher, dtype: int64
False    8287
Name: Title, dtype: int64
False    6509
True     1778
Name: Author, dtype: int64
False    8287
Name: Contributors, dtype: int64
False    8287
Name: Issuance type, dtype: int64
False    8287
Name: Flickr URL, dtype: int64


In [39]:
df[df['Publisher'].isnull()][['Place of Publication','Date of Publication']]

Place of Publication  Date of Publication
7                  Oxford                  NaN
8                  London               1676.0
9                  London               1679.0
11            Christiania               1859.0
12                Firenze               1888.0
13              Amsterdam               1839.0
14                 Savona               1897.0
16                  Paris               1860.0
17                  Paris               1873.0
18            Puerto-Rico               1866.0
26                 Milano               1893.0
29                   Wien               1896.0
34                Leipzig               1894.0
35               Plymouth                  NaN
39                   1845               1846.0
46                Upsaliæ               1833.0
49                Yverdon               1777.0
51         St. Petersburg               1827.0
52                 Berlin               1866.0
57                 Sceaux               1883.0
59                Coblenz               1865.0
61                Leipzig               1853.0
64    Carlstad, Stockholm               1852.0
65               Carlstad               1857.0
67                  Paris               1869.0
75             Allahabad]               1880.0
77              Stuttgart               1887.0
81        Санктпетербургъ               1889.0
93                  Paris               1896.0
96                 Napoli               1821.0
...                   ...                  ...
8237    Merida de Yucatan               1846.0
8238         Buenos Aires               1890.0
8240                Paris               1879.0
8241                Paris               1880.0
8242         Москва, 1860                  NaN
8243               Berlin               1841.0
8244            Stuttgart               1847.0
8245               London                  NaN
8246               Moskwa               1835.0
8247               London               1803.0
8248              Leipzig               1889.0
8249              Leipzig               1856.0
8250              Zürich               1888.0
8252               London               1869.0
8253               London               1831.0
8254                 Wien               1873.0
8256               London                  NaN
8257               London                  NaN
8259               London                  NaN
8260               London                  NaN
8262               London               1879.0
8263               London                  NaN
8264             Pontoise               1889.0
8267              Jackson                  NaN
8268               London                  NaN
8269               London               1708.0
8270               London               1868.0
8272               London                  NaN
8282               London               1838.0
8286               London               1834.0

[4195 rows x 2 columns]

# Highest no. of nan's are in 'Publisher' column. 
## Now to check How many nans we have in 'Date of publication' col, where 'Publisher' has nan's

In [40]:
nan_check = df[df['Publisher'].isnull()][['Place of Publication','Date of Publication','Identifier']]

In [41]:
nan_check.head()

Place of Publication  Date of Publication  Identifier
7                Oxford                  NaN         667
8                London               1676.0         874
9                London               1679.0        1143
11          Christiania               1859.0        1808
12              Firenze               1888.0        1905

In [42]:
nan_check['Date of Publication'].isnull().value_counts()

False    4013
True      182
Name: Date of Publication, dtype: int64

In [43]:
nan_check['Identifier'].isnull().value_counts()

False    4195
Name: Identifier, dtype: int64

In [44]:
nan_check['Place of Publication'].isnull().value_counts()

False    4183
True       12
Name: Place of Publication, dtype: int64

# We see that 182 out of 192 nan's in 'Date of Publication' are where 'Publisher' column has nan's. Therefore it is safe to delete those nan's.

## Create a new df ommiting all the null values in 'Publisher' column.

In [45]:
cdf = df[df['Publisher'].notnull()]

In [46]:
cdf['Publisher'].isnull().value_counts()

False    4092
Name: Publisher, dtype: int64

In [47]:
for col in cdf.columns:
    print(cdf[col].isnull().value_counts())

False    4092
Name: Identifier, dtype: int64
False    4092
Name: Place of Publication, dtype: int64
False    4082
True       10
Name: Date of Publication, dtype: int64
False    4092
Name: Publisher, dtype: int64
False    4092
Name: Title, dtype: int64
False    3064
True     1028
Name: Author, dtype: int64
False    4092
Name: Contributors, dtype: int64
False    4092
Name: Issuance type, dtype: int64
False    4092
Name: Flickr URL, dtype: int64


# There are 1028 nan's in 'Author' column, It would be better to delete the col rather than deleting just the nan's. And then delete the remaining nan's from 'Date of Publication' col.

In [48]:
cdf.drop('Author',axis=1, inplace=True)

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [49]:
cdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4092 entries, 0 to 8285
Data columns (total 8 columns):
Identifier              4092 non-null int64
Place of Publication    4092 non-null object
Date of Publication     4082 non-null float64
Publisher               4092 non-null object
Title                   4092 non-null object
Contributors            4092 non-null object
Issuance type           4092 non-null object
Flickr URL              4092 non-null object
dtypes: float64(1), int64(1), object(6)
memory usage: 287.7+ KB


In [50]:
cdf.dropna(axis=0, how='any', inplace=True)

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [51]:
for col in cdf.columns:
    print(cdf[col].isnull().value_counts())

False    4082
Name: Identifier, dtype: int64
False    4082
Name: Place of Publication, dtype: int64
False    4082
Name: Date of Publication, dtype: int64
False    4082
Name: Publisher, dtype: int64
False    4082
Name: Title, dtype: int64
False    4082
Name: Contributors, dtype: int64
False    4082
Name: Issuance type, dtype: int64
False    4082
Name: Flickr URL, dtype: int64


In [52]:
cdf.head(50)

Identifier Place of Publication  Date of Publication  \
0          206               London               1879.0   
1          216               London               1868.0   
2          218               London               1869.0   
3          472               London               1851.0   
4          480               London               1857.0   
5          481               London               1875.0   
6          519               London               1872.0   
10        1280             Coventry               1802.0   
15        2854               London               1865.0   
19        3131             New York               1899.0   
20        4598                 Hull               1814.0   
21        4884               London               1820.0   
22        4976               Oxonii               1800.0   
23        5382               London               1847.0   
24        5385               London               1897.0   
25        5389               London               1897.0   
27        6036               London               1805.0   
28        6821             Aberdeen               1837.0   
30        7630             Abingdon               1898.0   
31        8239               Quebec               1899.0   
32        8435               London               1892.0   
33        8440               London               1894.0   
36       13074               London               1899.0   
37       13364               London               1885.0   
38       14466               London               1860.0   
40       15141               London               1893.0   
41       15146               London               1894.0   
42       16173                  enk               1817.0   
43       16543               London               1816.0   
44       16544               Boston               1817.0   
45       17602               London               1897.0   
47       18173               London               1804.0   
48       18218            Edinburgh               1859.0   
50       18469              Londini               1799.0   
53       20960               London               1853.0   
54       21119               London               1874.0   
55       21430               London               1790.0   
56       21782               London               1817.0   
58       22105               London               1795.0   
60       23565               London               1877.0   
62       24878               London               1886.0   
63       27797               London               1834.0   
66       28758               Boston               1866.0   
68       28873               London               1853.0   
69       28876               London               1875.0   
70       29031               London               1827.0   
71       29032               London               1828.0   
72       29285               London               1876.0   
73       29505            A. Millar               1758.0   
74       29844             New York               1846.0   

                                            Publisher  \
0                                    S. Tinsley & Co.   
1                                        Virtue & Co.   
2                               Bradbury, Evans & Co.   
3                                       James Darling   
4                                Wertheim & Macintosh   
5                                   William Macintosh   
6                                          The Author   
10                               Printed by J. Turner   
15                                     E. Moxon & Co.   
19                                          W. Abbatt   
20                                         The Author   
21                                  J. Hatchard & Son   
22                                     J. Cooke, etc.   
23                                       Punch Office   
24                              Bradbury, Agnew & Co.   
25                             